### ISSUE TITLE
Incorrect behavior of Chromium web store v.1.5.3

### ISSUE THREAD

**Issue by Rusenche** [NeverDecaf/chromium-web-store#110]
Incorrect behavior of Chromium web store v.1.5.3

Hello.

I will try to explain clearly and in detail what I mean by incorrect and where the problem is!

I am using Ungoogled Chromium v.107.0.5304.88 x64.

Just to clarify, I switched to Chromium web store v.1.5.3 5-10 days ago.

While using the specified browser, Chromium web store v.1.5.3 does not show me any available update of any extension.

I am attaching a photo: https://i.postimg.cc/brLZTFzw/1.png

I was amazed how none of my extensions could go 5-10 days without an update.

And I just clicked on the Chromium web store icon. And it turns out I have a total of 4 extensions with updates available!

I am attaching a photo: https://i.postimg.cc/kXF0jP3c/2.png

Just FYI: While using Chromium web store v.1.4.4 - it always showed correctly that updates were available when there were, without having to click the icon until now!

Can you please fix this because there is no proper preview setting when icon extension updates are available?! Because as you understand - if I don't click on the Chromium web store v.1.5.3 icon - it doesn't notify me with a preview (numbers that show how many extensions have available updates). However - look at picture 1 and then look at picture 2. There is something wrong.

**Comment by NeverDecaf** [comment]
Can you first check to make sure you haven't disabled automatic updates in the options? 
![image](https://user-images.githubusercontent.com/19150416/200140668-258c055c-33b5-4795-9a9e-91ca86db42c2.png)
If these boxes are not both checked then cws will *only* check for updates when you click on the badge.

**Comment by Rusenche** [comment]
> [NeverDecaf](https://github.com/NeverDecaf): Can you first check to make sure you haven't disabled automatic updates in the options? ![image](https://user-images.githubusercontent.com/19150416/200140668-258c055c-33b5-4795-9a9e-91ca86db42c2.png) If these boxes are not both checked then cws will _only_ check for updates when you click on the badge.

Of course, this assumption would not be possible just because I wrote in the previous opinion that I had no problem with Chromium web store v.1.4.4. (There's no way even logically, if both checkboxes are unchecked in Chromium web store v.1.4.4 - it works perfectly.)

But anyway, here's a pic of what's tagged: https://i.postimg.cc/9QTG9mJP/sas.png

There is something wrong somewhere and I don't seem to want to go back to the perfectly working version (v.1.4.4) at least for now...

I am ready to help if needed (suggest where to look or where to test to find the problem) if there is a way to see where there is an error or why the icon must be clicked to visualize that they are available updates.

**Comment by NeverDecaf** [comment]
To give you some context: Version 1.5+ uses Manifest v3. One of the major changes in Manifest v3 is the removal of persistent background scripts. Manifest v3 extensions are forced to use [service workers](https://developer.chrome.com/docs/extensions/mv3/migrating_to_service_workers/) instead. One limitation of service workers is that they cannot run persistently in the background and will be terminated after ~1 minute (IIRC) of inactivity. The automatic update check is done in a background script (service worker in mv3). This check is done via an [alarm](https://developer.chrome.com/docs/extensions/reference/alarms/) which should, in theory, start the inactive service worker and update the badge. In previous versions (1.4.4) the persistent background script would listen for this alarm event while **still running** in the background.

Why this doesn't work in your setup I don't know, as it works perfectly fine in mine (ungoogled chromium 107.0.5304.88). Furthermore, the code for the update check via chrome.alarms has not been changed from version 1.4.4 to 1.5.3. My best guess is that there is simply some issue with Manifest v3/your browser which is causing the alarm event to either not fire, or not start the service worker.

I'm not sure how I can help with debugging this as I cannot reproduce this issue. One more thing you can try is updating your browser to the latest version (if it isn't already) as each version receives many bug fixes affecting manifest v3 extensions. I would honestly recommend switching back to 1.4.4 as long as that is an option. Manifest v3 has been consistently buggy with features that were working one day being broken the next; I only switched because Google repeatedly threatens to remove support for v2 while simultaneously extending the deadline for this deprecation. [This page](https://developer.chrome.com/docs/extensions/mv3/mv2-sunset/) seems to change every time I visit it, with the deadline moving forward a few months/years each time.

**Comment by kinoegit** [comment]
Thank you for your great work NeverDecaf!
Same issue seems to happen here with v.107.0.5304.87. I downgraded Chromium Web Store to v1.4.4 but i didn't feel good with downgrading to Manifest v2. Subsequently reinstalling Chromium Web Store (to 1.5.3) and various extensions seems to do the job for now, will see

**Comment by Rusenche** [comment]
> [NeverDecaf](https://github.com/NeverDecaf): as it works perfectly fine in mine (ungoogled chromium 107.0.5304.88). One more thing you can try is updating your browser to the latest version (if it isn't already).
> Edit: Should have asked this in my first reply but is there any error output in `chrome:extensions`? It's possible the error is unrelated to the alarm and caused by something else entirely.

First comment on this topic:

> [Rusenche](https://github.com/Rusenche): I am using Ungoogled Chromium v.107.0.5304.88 x64.

:}

"chrome:extensions" - no, there is no error that shows up or says an error is present.

I removed version 1.5.3.

Via "crx" I installed version 1.4.4. No problem - it instantly shows that there are already 6 extensions available with updates - one update of which is "Chromium web store" v.1.5.3. I clicked to install the update on just this extension - it gave an error saying it's not available (because it's looking for it on the desktop?!) "file:///C:/Users/%D0%9E%D0%B3%D0%B8% D1%82%D0%BE/Desktop/Chromium.Web.Store.crx" -> this is written when clicking to install only this update from the available 6. 

"No access to the file". "ERR_FILE_NOT_FOUND".

I am attaching a photo: https://i.postimg.cc/hj3qv3gw/aaaaa.png

And despite the error - after deliberately restarting Ungoogled Chromium v.107.0.5304.88 x64 version 1.5.3. it had installed as it had given an error - see above in the picture. (facepalm)

I manually downloaded the "Chromium web store" v.1.5.3 as a "crx" file and installed it manually (it still gave an error that it couldn't just download from the site from here). Still not showing available updates. Then I removed it. Then without installing its previous version, I directly installed the latest updated version 1.5.3 - via "crx" file. It is being installed. And it still doesn't show any available updates. Yet apparently there is a problem after version 1.4.4.

**Comment by NeverDecaf** [comment]
Thank you for testing all this. Can I first ask whether you have the `chrome://flags/#extension-mime-request-handling` flag set? Inability to install by direct download is usually caused by the lack of this flag. Additionally, are you using your own build of ungoogled chromium or one of the community builds? I will attempt to replicate your steps and see if I encounter the same issue but I need to be sure I'm using the exact same browser as you are.

As for the ERR_FILE_NOT_FOUND error, this is caused by a workaround used in versions prior to 1.5. I can't remember the specifics but the gist of it is that the file is first downloaded, then opened via a local `file:///` URI as otherwise you would get the error: `Apps, extensions and user scripts cannot be added from this website`. The reason you see file not found errors is that chromium immediately deletes CRX files when they are installed (or fail to install), meaning your browser *should* have already handled the CRX install (but evidently didn't) causing the file to no longer exist. Regardless, this is a separate issue and doesn't exist in 1.5+.

**Comment by NeverDecaf** [comment]
> I installed version 1.4.4. No problem - it instantly shows that there are already 6 extensions available with updates

This is actually the result of a **bug** in versions <1.5 which caused chromium-web-store to *always* check for updates on startup, regardless of user settings. 1.5+ correctly respects the user's setting and, if automatic updates is not checked, will **never** check for updates unless the badge is clicked. This makes me wonder if your automatic update checking (every X minutes) may not be working even in 1.4.4, and instead it was simply checking for updates every time you restarted your browser. Either way, to correctly test if this is working in 1.5.3 after a fresh install you will need to either wait 60 minutes or set the update time to 1 minute and then wait 1 minute to see if the badge changes.

**Comment by NeverDecaf** [comment]
@kinoegit Thank you for the info and please keep us updated on whether background update checking works for you after the reinstall. Can you also confirm whether you are using your own build of ungoogled chromium or one of the community builds ([marmaduke's](https://github.com/macchrome/winchrome/) for example)?

**Comment by Rusenche** [comment]
> [NeverDecaf](https://github.com/NeverDecaf) Thank you for testing all this. Can I first ask whether you have the `chrome://flags/#extension-mime-request-handling` flag set? Inability to install by direct download is usually caused by the lack of this flag. Additionally, are you using your own build of ungoogled chromium or one of the community builds? I will attempt to replicate your steps and see if I encounter the same issue but I need to be sure I'm using the exact same browser as you are.
> 
> As for the ERR_FILE_NOT_FOUND error, this is caused by a workaround used in versions prior to 1.5. I can't remember the specifics but the gist of it is that the file is first downloaded, then opened via a local `file:///` URI as otherwise you would get the error: `Apps, extensions and user scripts cannot be added from this website`. The reason you see file not found errors is that chromium immediately deletes CRX files when they are installed (or fail to install), meaning your browser _should_ have already handled the CRX install (but evidently didn't) causing the file to no longer exist. Regardless, this is a separate issue and doesn't exist in 1.5+.

> Either way, to correctly test if this is working in 1.5.3 after a fresh install you will need to either wait 60 minutes or set the update time to 1 minute and then wait 1 minute to see if the badge changes.

1. https://i.postimg.cc/jjFY0GFL/extension-mime-request-handling.png
2. I use: https://github.com/macchrome/winchrome/releases/tag/v107.5304.88-M107.0.5304.88-r1047731-Win64
3. Third comment from this thread:
> [Rusenche](https://github.com/Rusenche): But anyway, here's a pic of what's tagged: https://i.postimg.cc/9QTG9mJP/sas.png - > It is set to 1 minute. :}

It is set to 1 minute. :}

I have provided the information about the ideas well in advance  you to find and solve the problem. :P

4. It's not instant, but it's like ~1 minute - it's instinctive, I haven't detected it. But with version 1.5.3 this is not visualized even after hours, nor is the browser restarted intentionally, and I have been using it for hours.

**Comment by NeverDecaf** [comment]
Thank you for all the info, unfortunately we seem to have identical setups so I don't know what else I can do to try and reproduce this problem. The only other thing I can think of is maybe, somehow, one of your other extensions is interfering with the update checking process, though I'm not sure if that is even possible. Since that seems to be the only difference between our browsers maybe you can try disabling all extensions except 1-2 that need updates and see if that does something? Beyond that I'm at a complete loss as to what to try.

**Comment by Rusenche** [comment]
> [NeverDecaf](https://github.com/NeverDecaf): Благодаря ви за цялата информация, за съжаление изглежда, че имаме идентични настройки, така че не знам какво друго мога да направя, за да опитам да възпроизведа този проблем. Единственото друго нещо, за което се сещам, е може би някое от другите ви разширения да се намесва в процеса на проверка на актуализацията, въпреки че не съм сигурен дали това изобщо е възможно. Тъй като това изглежда е единствената разлика между нашите браузъри, може би можете да опитате да деактивирате всички разширения с изключение на 1-2, които се нуждаят от актуализации, и да видите дали това ще помогне? Освен това съм в пълна загуба какво да опитам.


Logically, this would be close to impossible, because another extension according to your thesis would somehow block version 1.5.3, but not block version 1.4.4.

I have one other option left - to download the portable version of the browser, the link of which I provided in my above comment and... I'm still considering how to get old versions of some extensions to conduct the experiment with version 1.5.3 if it will respond adequately as version 1.4.4.

(I almost rejected the idea of ​​using all my current - "User Data" - settings and transferring them to the new portable version, and from there removing many extensions to leave only a few - like 5 or 10, but there is no guarantee that it won't affect any specific setting in chrome://flags/ either. Because - unfortunately - there is no possibility for the settings from "chrome://flags/" - to be extracted/exported in a separate file only those...). These are my thoughts right now.

**Comment by NeverDecaf** [comment]
An easy way to test is to modify the `manifest.json` of chromium web store itself (and install it via "Load Unpacked"), just change the "version" value to something lower. This isn't a perfect test since the update check process is slightly different for extensions that aren't listed in the Chrome web store, but should suffice for this specific issue.

**Comment by kinoegit** [comment]
> @kinoegit Thank you for the info and please keep us updated on whether background update checking works for you after the reinstall. Can you also confirm whether you are using your own build of ungoogled chromium or one of the community builds ([marmaduke's](https://github.com/macchrome/winchrome/) for example)?

I'm using the 'official' [AUR package](https://github.com/ungoogled-software/ungoogled-chromium). Will keep you updated wether background update checking works.

**Comment by Rusenche** [comment]
Test with a new portable version of Ungoogled Chromium v.107.0.5304.88 x64 with absolutely no personal settings ported!

The results are really strange!

1. Chromium web store version 1.5.3 displays correctly that there is an extension update.
2. The strangest thing is that out of 4 extensions installed, only Chromium web store 1.5.3 is an up-to-date version, the other three extensions are installed from github "crx" files with very old versions. Why is it weird?! Because Chromium web store 1.5.3 registers that there is only 1 update available, which is flat out wrong.

I am attaching a photo: https://i.postimg.cc/t4cdrqBM/sas.png

I mean after installing very old versions via "crx" from github - Chromium web store 1.5.3 is set to check for updates every 1 minute: https://i.postimg.cc/nh99hv13/sas-2.png and yet Chromium web store 1.5.3 finds only 1 available update: https://i.postimg.cc/tRhjp7PF/sas-3.png

And despite the results - I'm not going to dump my personal settings in my Ungoogled Chromium v.107.0.5304.88 x64 profile.

In other words, the results are surprising and contradictory.

**Comment by NeverDecaf** [comment]
I believe this behavior is caused by 2/3 of those extensions not being packed with a [key file](https://developer.chrome.com/docs/extensions/mv3/manifest/key/). If you obtained them from github instead of the chrome web store it isn't required that the crx be packaged with a key file (or packaged at all), meaning the id used to check for updates would not match the id on the chrome web store, causing all update checks to fail. You can verify this by locating the `manifest.json` of those installed extensions in your user profile directory `%LOCALAPPDATA%\Chromium\User Data\Default\Extensions` and checking if the manfiest contains a "key". I tested 2 of those extensions myself and ClearURLs does while IDCAC does not, which is in line with what you experienced.

**Comment by Rusenche** [comment]
Clarification that I explained myself an hour ago: Only the Clearurls extension on github was uploaded as "crx". All other extensions were uploaded to github as "zip" - I unzipped them accordingly - chrome://extensions/ - here I loaded the unzipped and the extension appeared. Then I packaged the extension and got two files. For example:
1. IDCAC-chrome.pem;
2. IDCAC-chrome.crx.
I then did this for the rest of the extensions that were uploaded to github as "zip" files, removed the loaded bundles and loaded the newly created "crx" ones - three such extensions were added this way. This led me to think that the Chromium web store v.1.5.3 is not finding the way to update them.



**New test:**

After downloading a portable version from https://github.com/macchrome/winchrome/releases/tag/v107.5304.88-M107.0.5304.88-r1047731-Win64 - I added my personal settings - "User Data".
1. I disabled all my extensions leaving only Chromium web store v.1.5.3. No updates showed up, although I restarted the browser and waited a minute to be sure - it didn't show any extension updates available.
2. In chrome://flags/ - I reset all settings. I restarted the browser. Again, the search result was missing.
3. In chrome://settings/reset - I reset the settings to default settings. I restarted the browser. Again, the desired result was not there.

I have no other ideas. In fact, there is nothing else to test, everything possible was tested.

It's amazing how neither disabling extensions, nor resetting flags, nor resetting browser-wide settings in any way caused the Chromium web store v.1.5.3 to show that there are updates available to my available extensions.

I reset and reset all possible settings.

I'm amazed none of this worked.

It turns out that Chromium web store v.1.4.4 is indispensable until we find what the problem is in version 1.5.3.

**Comment by NeverDecaf** [comment]
By packing extensions yourself you are changing the id associated with them, which is why the .pem file is generated. The only way to pack an extension so that it will have the same id as the web store version is to pack it with the author's private key file (.pem), which they should never share with anyone. This is why you do not see updates for any extensions you packed yourself, they will never have a matching id with the web store version. You can verify this by finding the web store page for any extension you have installed and comparing the id in the url to the one shown in `chrome://extensions`. For example, in your above screenshot IDCAC shows an id of `ekejoopfgpekkbgndaonnnoeogcflchh` but on the web store the id is [fihnjjcciajhdojfnbdddfaoknhalnja](https://chrome.google.com/webstore/detail/i-dont-care-about-cookies/fihnjjcciajhdojfnbdddfaoknhalnja), as you can see from the url. For all extensions from the chrome web store, the id is used to check for updates and therefore must match for the check to succeed.

As a final test, can you try installing chrome web store 1.5.2.2 and see if it shows that an update is available?

**Comment by Rusenche** [comment]
Very important clarification: I thought of one last possible elementary test.

I tried for test @teeminus - ungoogled-chromium_107.0.5304.88-1.1_windows_**x64**.zip and ungoogled-chromium_107.0.5304.88-1.1_windows_**86**.zip.

Accordingly, I transferred the user folder with the settings - "User Data".

When running @teeminus' browser compilation/build in Chromium web store extensions installed it was v.1.4.4. Since I have version 1.5.3 of "crx", I just dragged the file into chrome://extensions/ and instantly the current version in @teeminus' browser became 1.5.3.

I've restarted @teeminus's browser several times. No matter how many times I restarted it - every time it displayed that updates were available to my installed extensions.

Bottom line: The problem turned out to be @marmaduk's
 - https://github.com/macchrome/winchrome/releases/tag/v107.5304.88-M107.0.5304.88-r1047731-Win64

(As far as I know his attitude towards users, if a topic is posted in his repository - either there will be no response and the topic will be closed, or his advice will be to use another browser. These are my observations of the grumpiness of his character!)

Somewhere something is broken in his build, I don't know if it's the latest one or how many builds of it back, but I've tested that something in his build breaks the rendering of available updates on the Chromium web store was v.1.5.3. This was confirmed by my test experiments, if for another or others it is not confirmed - I don't know, for myself I proved that in this case @teeminus' build worked flawlessly against the Chromium web store was v.1.5.3.

**Comment by Rusenche** [comment]
> [NeverDecaf](https://github.com/NeverDecaf): As a final test, can you try installing chromium web store 1.5.2.2 and see if it shows that an update is available?

I can confirm that Chromium web store v.1.5.2.2 in @mamraduk's version (Ungoogled Chromium v.107.0.5304.88 x64) works flawlessly. Apparently you found the cause of the problem?!

**Comment by NeverDecaf** [comment]
Does it find updates for every extension or just for itself?

**Comment by Rusenche** [comment]
Chromium web store v.1.5.2.2 finds all available extension updates. That's why I wrote in the previous comment that it also works flawlessly on the @marmaduk version (Ungoogled Chromium v.107.0.5304.88 x64).

![Ahh](https://user-images.githubusercontent.com/18578128/200179099-469d93fb-79f5-484c-8cca-827f282541cd.png)

**Comment by NeverDecaf** [comment]
There are effectively 0 differences between 1.5.2.2 and 1.5.3 with regards to update checks so I can't imagine how this is possible.

**Comment by Rusenche** [comment]
If you claim it - I also cannot analyze how it is possible.

But as a tip, if others have the same problem as me - use the builds of @teenminus who doesn't have this problem.

@NeverDecaf, did you read https://github.com/NeverDecaf/chromium-web-store/issues/110#issuecomment-1304811332?!

**Comment by NeverDecaf** [comment]
I did but, again, I am using the exact marmaduke build you are and am not experiencing this issue, so I don't want to jump to any conclusions. I will, of course, leave this issue open so others who experience this problem can try either fix for themselves.

**Comment by Rusenche** [comment]
@NeverDecaf, I offer you the best option - I apply my personal settings (of course - no personal passwords and no bookmarks) to look at the problem. In my settings folder I intentionally left Chromium web store v.1.4.4. Add version 1.5.3 yourself and observe/investigate.

https://krakenfiles.com/view/s9nEkZkouD/file.html

Password: [NeverDecaf](https://github.com/NeverDecaf)

**Comment by NeverDecaf** [comment]
Here are the exact steps I've taken:
1. Delete my own `User Data` directory and replace it with yours.
2. Download and launch ungoogled chromium portable from the above [links](https://github.com/macchrome/winchrome/releases/tag/v107.5304.88-M107.0.5304.88-r1047731-Win64)
3. Visit `chrome://extensions` and see that no extensions are shown.
4. Navigate to `User Data\Default\Extensions` and see that all files and folders (except one .pem file) have been deleted.
5. Install chromium web store 1.5.3 by drag-and-drop
6. Install [clearurls-1.20.0-chrome.crx](https://github.com/ClearURLs/Addon/releases) by drag-and-drop
7. Verify that chromium web store shows an update is available:
![image](https://user-images.githubusercontent.com/19150416/200198221-ce04b89d-9e65-4b77-8d8f-0186be76cedf.png)

I should note that your settings for chromium web store were saved despite the extension itself being deleted (when I checked extension options the update time was already set to 1 minute.) I don't know why chromium decided to invalidate all your installed extensions when first launched but besides that everything seems to be working. I also checked `chrome://flags` to see if your other settings were preserved and I do indeed see all the flags that you have changed. I then visited `chrome://settings` and was greeted by this:
![image](https://user-images.githubusercontent.com/19150416/200198434-5220634a-7dc3-4a75-a555-817fbb817c18.png)

**Comment by Rusenche** [comment]
Strange. Even after reinstalling the Windows OS, saving the "User Data" folder and even moving it to another computer has always been without that "defect" you described, and works flawlessly.

Besides asking you to do a second experiment under the same conditions?! Something somewhere prevented almost all of my extensions from disappearing from being active for you too.

**Comment by NeverDecaf** [comment]
I did try multiple times but every time I got the same result: the contents of the `Extensions` folder was automatically deleted shortly after launching chromium. You said you removed your saved passwords and bookmarks, perhaps something about this process caused the profile to become corrupted.

**Comment by Rusenche** [comment]
After a lot of testing on @marmaduk and @teeminus builds/compilations I found the problem.

Categorically, the problem has been identified and resolved.

Finding: It turned out 100% that the teeminus build also had the problem discussed in this thread, but it took many browser restarts and many browser waits for me to "detect/catch" the problem.

My apologies to @marmaduk!

@NeverDecaf, here are the steps to investigate (if you decide to reproduce my exact problem) that you should experience this issue discussed in this thread:

1. Run Ungoogled Chromium v.107.0.5304.88 x64.
2. Close Ungoogled Chromium v.107.0.5304.88 x64.
3. Delete the entire folder: Service Worker. Located in %Profile%\Default\Service Worker.
4. Run Ungoogled Chromium v.107.0.5304.88 x64.

@NeverDecaf, you will notice yourself that Chromium web store v.1.5.3 does not display that there are updates available to your installed extensions.


Conclusion: The problem was in deleting the entire "Service Worker" folder. It's strange to me why this problem appeared with Chromium web store v.1.5.3 and it doesn't exist with Chromium web store v.1.5.2.2, with the same "Service Worker" folder deleted so far for me.

@NeverDecaf, maybe something from the new version of Chromium web store v.1.5.3 is being redirected and/or some files from your new version are contained in "Service Worker" and deleting that folder causes the issue discussed in this thread.

I confirm that it is safe to delete the "CacheStorage" folder. Located in "%Profile%\Default\Service Worker\CacheStorage".


Last but not least: @NeverDecaf, thank you for responding, thank you for being understanding, empathetic and respectful. You also have my respect.

**Comment by NeverDecaf** [comment]
Glad you figured it out. As mentioned above all manifest v3 extensions are [required](https://developer.chrome.com/docs/extensions/mv3/migrating_to_service_workers/) to use service workers instead of background pages, so deleting the `Service Worker` directory may also break other manifest v3 extensions, not just this one.

**Comment by kinoegit** [comment]
background update checking is working, though after updating the small red counter button disappears only after a restart of the browser

**Comment by NeverDecaf** [comment]
@kinoegit Does it reappear again after a while (after the first update check)?

**Comment by kinoegit** [comment]
@NeverDecaf i don't quite get what you mean- sorry

**Comment by NeverDecaf** [comment]
Ah, I just understood what you meant. You're experiencing #4: the extension can not (or will not) update while it is running, so it only updates when you restart your browser.

**Comment by kinoegit** [comment]
No no, I meant everything is working as expected

**Comment by NeverDecaf** [comment]
OK, sounds good. Thank you for keeping us updated. I'll close this issue since it has been resolved.
